spanconfigsqltranslator,jobsprotectedts: add `ignore_if_excluded_from_backup` to SpanConfig

adityamaru opened the pull request:
This change is a follow up to https://github.com/cockroachdb/cockroach/pull/75451 which taught ExportRequests
to noop on ranges marked as exclude_data_from_backup. This change

This diff does two things:

- It adds an `ignore_if_excluded_from_backup` bit to ptpb.Target that is set
on PTS records written by backup schedules and jobs.

- It adds an `ignore_if_excluded_from_backup` bit to the ProtectionPolicy that
is shipped to KV as part of the SpanConfig.

In a follow up PR, this bit on the SpanConfig will be used in conjunction with
`exclude_data_from_backup` to decide whether or not to ignore the ProtectionPolicy
when making GC decisions on a span. All other consumers of PTS records will
default to setting this bit to false, and so their ProtectionPolicies will always
influence GC even if `exclude_data_from_backup` is set to true.

Informs: https://github.com/cockroachdb/cockroach/issues/73536

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76831)
<!-- Reviewable:end -->

dt commented on pkg/spanconfig/spanconfigsqltranslator/protectedts_state_reader.go:
```diff
@@ -70,22 +75,32 @@ func (p *protectedTimestampStateReader) getProtectionPoliciesForSchemaObject(
 	for _, record := range ptsState.Records {
+		// TODO(adityamaru): When schedules start writing protected timestamp
+		// records extend this logic to also include those records.
+		isBackupProtectedTimestampRecord, err := jobsprotectedts.IsRecordWrittenByJobType(ctx, jr, txn,
```
Huh, so recently I haven't followed everything here all that closely, so maybe this is a silly question, but it seems weird to me to go by who created the pts record rather than just have that creator set a bit in the record it creates and then look at that bit?

dt replied:
I don't really like looking up the job here much. Is there not a proto in that table we can just add a field to vs a whole new column?

adityamaru replied:
I could stick it in https://github.com/cockroachdb/cockroach/blob/5a18c3ab8f4d1d096bc3a2d98bbe712d547c9cd1/pkg/kv/kvserver/protectedts/ptpb/protectedts.proto#L140. I guess it sorta fits that its information needed when evaluating the pts record and what it targets.

adityamaru replied:
We don't write the Record as a proto unfortunately, we pick out its fields.

```
   id        UUID NOT NULL,
   ts        DECIMAL NOT NULL,
   meta_type STRING NOT NULL,
   meta      BYTES,
   num_spans INT8 NOT NULL, -- num spans is important to know how to decode spans
   spans     BYTES NOT NULL,
   verified  BOOL NOT NULL DEFAULT (false),
   target    BYTES,         -- target is an encoded protobuf that specifies what the pts record will protect
```

adityamaru replied:
I think I like the idea of sticking it into `ptpb.Target`. Will rework this PR to do that instead.

adityamaru committed:
spanconfigsqltranslator,jobsprotectedts: add ignore_if_excluded_from_backup to SpanConfig

This change is a follow up to #75451 which taught ExportRequests
to noop on ranges marked as exclude_data_from_backup. This change

This diff does two things:

- It adds an `ignore_if_excluded_from_backup` bit to ptpb.Target that is set
on PTS records written by backup schedules and jobs.

- It adds an `ignore_if_excluded_from_backup` bit to the ProtectionPolicy that
is shipped to KV as part of the SpanConfig.

In a follow up PR, this bit on the SpanConfig will be used in conjunction with
`exclude_data_from_backup` to decide whether or not to ignore the ProtectionPolicy
when making GC decisions on a span. All other consumers of PTS records will
default to setting this bit to false, and so their ProtectionPolicies will always
influence GC even if `exclude_data_from_backup` is set to true.

Informs: #73536

Release note: None

dt reviewed (approved):
nice, I like this approach

adityamaru commented:
@irfansharif do you want to have a quick look at this? Most of the spanconfig changes are to how we print DD test output.

adityamaru commented:
TFTRs!

bors r=dt,irfansharif
